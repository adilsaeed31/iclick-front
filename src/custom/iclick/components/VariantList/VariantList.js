import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import BadgeOverlay from "@reactioncommerce/components/BadgeOverlay/v1";
import { badgeStatus, BADGE_LABELS } from "@reactioncommerce/components/BadgeOverlay/v1/utils";
import InventoryStatus from "@reactioncommerce/components/InventoryStatus/v1";
import VariantItem from "custom/iclick/components/VariantItem";
import ProductDetailOptionsList from "custom/iclick/components/ProductDetailOptionsList";
import Divider from "custom/iclick/components/Divider";

export default class VariantList extends Component {
  static propTypes = {
    currencyCode: PropTypes.string.isRequired,
    onSelectOption: PropTypes.func,
    onSelectVariant: PropTypes.func,
    product: PropTypes.object.isRequired,
    selectedOptionId: PropTypes.string,
    selectedVariantId: PropTypes.string,
    variants: PropTypes.arrayOf(PropTypes.object).isRequired
  }

  renderVariant = (variant) => {
    const { currencyCode, onSelectVariant, selectedVariantId } = this.props;

    const active = selectedVariantId === variant._id;

    return (
      <div key={variant._id}>
        <VariantItem
          currencyCode={currencyCode}
          handleClick={() => {
            onSelectVariant(variant);
          }}
          isActive={active}
          variant={variant}
        />
        {this.renderBadges(variant)}
      </div>
    );
  }

  renderBadges(variant) {
    const status = badgeStatus(variant, BADGE_LABELS);

    if (!status) return null;

    return (
      <div>
        <BadgeOverlay
          product={variant}
          filterOnly={"MERCHANDISING"}
          shouldShowPrimaryOnly={true}
          label={status.label}
        />
      </div>
    );
  }

  renderInventoryStatusText() {
    const { selectedOptionId, selectedVariantId, variants } = this.props;
    const selectedVariant = variants.find((variant) => variant._id === selectedVariantId);

    if (selectedOptionId) {
      // Check to make sure the selected option is from this current page, and not left over from a previous page
      const options =
        selectedVariant && Array.isArray(selectedVariant.options) && selectedVariant.options.length
          ? selectedVariant.options
          : null;

      if (options) {
        const selectedOption = options.find((option) => option._id === selectedOptionId);
        return <InventoryStatus product={selectedOption} />;
      }
    }

    // If we don't have an option, use the variant for inventory status information
    if (selectedVariantId) {
      return <InventoryStatus product={selectedVariant} />;
    }

    // We should always have a selected option or variant, so we should never get this far
    return null;
  }

  renderOptionsList() {
    const { onSelectOption, product, selectedOptionId, selectedVariantId, variants } = this.props;
    const selectedVariant = variants.find((variant) => variant._id === selectedVariantId);

    // If currently selected variant has options, then render them.
    const options =
      selectedVariant && Array.isArray(selectedVariant.options) && selectedVariant.options.length
        ? selectedVariant.options
        : null;

    if (!options) return null;

    return (
      <Fragment>
        <Divider label="Available Options" />
        <ProductDetailOptionsList
          productSlug={product.slug}
          onSelectOption={onSelectOption}
          options={options}
          selectedOptionId={selectedOptionId}
        />
      </Fragment>
    );
  }

  render() {
    const { variants } = this.props;
    return (
      <div>
        {variants.map(this.renderVariant)}
        {this.renderOptionsList()}
        <div>{this.renderInventoryStatusText()}</div>
      </div>
    );
  }
}
