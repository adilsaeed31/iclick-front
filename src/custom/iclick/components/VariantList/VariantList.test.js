import React from "react";
import renderer from "react-test-renderer";
import { MuiThemeProvider } from "@material-ui/core/styles";
import { Provider } from "mobx-react";

import catalogProduct from "components/ProductDetail/__mocks__/productData.mock";
import theme from "custom/reactionTheme";
import VariantList from "./VariantList";

const uiStore = {
  pdpSelectedOptionId: "cmVhY3Rpb24vY2F0YWxvZ1Byb2R1Y3RWYXJpYW50OkpobjJ0YzVaRkFzYVdiaXZR"
};

test("basic snapshot", () => {
  const component = renderer.create((
    <MuiThemeProvider theme={theme}>
      <Provider uiStore={uiStore}>
        <VariantList
          currencyCode="USD"
          product={catalogProduct}
          selectedOptionId="cmVhY3Rpb24vY2F0YWxvZ1Byb2R1Y3RWYXJpYW50OlNNcjRyaERGbll2Rk10RFRY"
          selectedVariantId="cmVhY3Rpb24vY2F0YWxvZ1Byb2R1Y3RWYXJpYW50OjZxaXFQd0JrZUpkdGRRYzRH"
          variants={catalogProduct.variants}
        />
      </Provider>
    </MuiThemeProvider>
  ));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
