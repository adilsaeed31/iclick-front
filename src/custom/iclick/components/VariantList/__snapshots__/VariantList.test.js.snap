// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
.c0 {
  position: relative;
  opacity: 0.5;
}

.c1 {
  position: relative;
}

.c2 {
  -webkit-font-smoothing: antialiased;
  color: #505558;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  font-size: 14px;
  font-style: normal;
  font-stretch: normal;
  font-weight: 400;
  -webkit-letter-spacing: .02em;
  -moz-letter-spacing: .02em;
  -ms-letter-spacing: .02em;
  letter-spacing: .02em;
  line-height: 1.25;
}

<div
  className="SkVariantList-variantsContainer-3"
>
  <div
    className="SkVariantList-variantItem-4"
  >
    <button
      className="MuiButtonBase-root-8 SkVariantItem-variantButton-5 SkVariantItem-activeVariant-6 SkVariantItem-soldOutVariant-7"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiTypography-root-11 MuiTypography-body1-20"
      >
        Sample Variant
      </span>
      <span
        className="MuiTypography-root-11 MuiTypography-body1-20"
      >
        $12.99 - $19.99
      </span>
    </button>
    <div
      className="SkVariantList-alert-1"
    >
      <div
        className="c0"
      />
    </div>
  </div>
  <div
    className="SkVariantList-variantItem-4"
  >
    <button
      className="MuiButtonBase-root-8 SkVariantItem-variantButton-5"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiTypography-root-11 MuiTypography-body1-20"
      >
        Sample Variant 2
      </span>
      <span
        className="MuiTypography-root-11 MuiTypography-body1-20"
      >
        $12.99 - $19.99
      </span>
    </button>
  </div>
  <div
    className="SkDivider-container-37"
  >
    <hr
      className="SkDivider-item-39"
    />
    <span
      className="MuiTypography-root-11 MuiTypography-body1-20 SkDivider-label-38"
    >
      Available Options
    </span>
    <hr
      className="SkDivider-item-39"
    />
  </div>
  <div
    className="MuiGrid-container-43 MuiGrid-spacing-xs-8-65 SkOptionsList-root-40"
  >
    <div
      className="MuiGrid-item-44"
    >
      <button
        className="MuiButtonBase-root-8 SkProductDetailOption-optionButton-140 SkProductDetailOption-isSelected-142"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiTypography-root-11 MuiTypography-body1-20 SkProductDetailOption-optionText-141"
        >
          Red
        </span>
      </button>
      <div
        className="SkOptionsList-alert-41"
      >
        <div
          className="c1"
        />
      </div>
    </div>
    <div
      className="MuiGrid-item-44"
    >
      <button
        className="MuiButtonBase-root-8 SkProductDetailOption-optionButton-140"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiTypography-root-11 MuiTypography-body1-20 SkProductDetailOption-optionText-141"
        >
          Green
        </span>
      </button>
    </div>
  </div>
  <div
    className="SkVariantList-inventoryStatus-2"
  >
    <div
      className="c2"
    >
      Backordered - ships when available
    </div>
  </div>
</div>
`;
